feat(ci): Support for install via CI (#135)
GoogleCloudPlatform/spinnaker-for-gcp: ci/CLOUD_BUILD.md

# Installing Spinnaker for GCP on Google Cloud Build

## Shared VPC

Please note that using Cloud Build to install Spinnaker for GCP with a Shared VPC host project is currently unsupported. For Shared VPC support, conduct the [setup](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp) in Cloud Shell.

## Service account

> Google Cloud Build has a default service account that can be found in the settings page of the Cloud Build UI. The Cloud Build service account must be configured with the following roles to install Spinnaker for GCP:

Reviewer: s/configured with/granted

> - Cloud Functions Developer - roles/cloudfunctions.developer

Author: It is a bit tricky because if the user is configuring through the UI they will be see the roles listed as the longform names, if they are adding them through the CLI they will need to add them using the role-paths

> - Cloud Functions Developer - roles/cloudfunctions.developer

Reviewer 2: Makes sense. Let it stand.

- Compute Network Viewer - roles/compute.networkViewer
- Kubernetes Engine Admin - roles/container.admin 
- Create Service Accounts - roles/iam.serviceAccountCreator
- Pub/Sub Editor - roles/pubsub.editor
- Cloud Memorystore Redis Admin - roles/redis.admin
- Service Usage Admin - roles/serviceusage.serviceUsageAdmin
- Source Repository Administrator - roles/source.admin
- Storage Admin - roles/storage.admin
- Project IAM Admin - roles/resourcemanager.projectIamAdmin
- Service Account User - roles/iam.serviceAccountUser

These roles can be enabled through the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).

## Properties file

In order to install Spinnaker for GCP, you must provide the installation script with a properties file. This properties file can be generated by executing [setup_properties.sh](../scripts/install/setup_properties.sh). Once you have a properties file, it must be provided to the installation script during the execution of the Cloud Build job.

## Submitting a Build

> Cloud Builds can be trigged using [gcloud](https://cloud.google.com/cloud-build/docs/running-builds/start-build-manually), [build triggers](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds), or [GitHub app triggers](https://cloud.google.com/cloud-build/docs/create-github-app-triggers). This repository provides the necessary files for installing Spinnaker for GCP with a gcloud triggered build. Follow these steps to start a build:

Reviewer: s/trigged/triggered

> 1. Create a new directory. The contents of this directory will be submitted to Cloud Build.

Author: Anywhere in the shell is ok as long as the user runs the `gcloud builds submit` from the newly created directory. Should I add a note indicating the command should be run from there?

`Submit the build to Cloud Build from within the directory you previously created: gcloud builds submit --timeout "25m"  --config cloudbuild.yaml .`

> 1. Create a new directory. The contents of this directory will be submitted to Cloud Build.

Reviewer 2: Yes, that looks good

# Using Cloud Build to install Spinnaker for GCP

## A note about Shared VPC support

You can't use Cloud Build to install Spinnaker for GCP with a shared VPC. For Shared VPC support, conduct the [setup in Cloud Shell](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp).

## Service account

To install Spinnaker for GCP, you need to grant the [Cloud Build service account](https://console.cloud.google.com/cloud-build/settings) the following roles:

- Cloud Functions Developer - roles/cloudfunctions.developer
- Compute Network Viewer - roles/compute.networkViewer
- Kubernetes Engine Admin - roles/container.admin 
- Create Service Accounts - roles/iam.serviceAccountCreator
- Pub/Sub Editor - roles/pubsub.editor
- Cloud Memorystore Redis Admin - roles/redis.admin
- Service Usage Admin - roles/serviceusage.serviceUsageAdmin
- Source Repository Administrator - roles/source.admin
- Storage Admin - roles/storage.admin
- Project IAM Admin - roles/resourcemanager.projectIamAdmin
- Service Account User - roles/iam.serviceAccountUser

> You can enable these roles using the IAM UI or with [gcloud](https://cloud.google.com/sdk/gcloud/reference/projects/add-iam-policy-binding).

Reviewer: s/enable/grant

## Enable Cloud Resource Manager API

> For Cloud Build to successfully retrieve IAM policies, you must enable the Cloud Resource Manager API. Visit this URL, substituting your project number.

Reviewer: Could give guidance on where to locate project number.

Reviewer: Actually, is this supposed to be "project id"?

Author: Originally, I used the project number to enable/disable but both seem to work. I will update to use project id and make this consistent with the rest of the documentation.

> The Cloud Resource Manager API must be enabled for Cloud Build to successfully retrieve IAM policies. Enable it for your project by visiting the below URL and substituting your project number.

Reviewer 2: ```suggestion
For Cloud Build to successfully retrieve IAM policies, you must enable the Cloud Resource Manager API. Visit this URL, substituting your project number.
```

## Submitting a Build

Cloud Builds can be triggered using [gcloud](https://cloud.google.com/cloud-build/docs/running-builds/start-build-manually), [build triggers](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds), or [GitHub app triggers](https://cloud.google.com/cloud-build/docs/create-github-app-triggers). The solution in this repository installs Spinnaker for GCP using a gcloud-triggered build. Follow these steps to start a build:

1. Create a new directory. The contents of this directory will be submitted to Cloud Build.
2. Place the generated properties file into that directory.
3. Copy the [cloudbuild.yaml](cloudbuild.yaml) file into the directory and edit the `user.name` and `user.email` used in the Git configuration steps.

```yaml
  - name: gcr.io/cloud-builders/git
    args: ['config', '--global', 'user.name', '<example-user>']
  - name: gcr.io/cloud-builders/git
    args: ['config', '--global', 'user.email', '<example-user@example.com>']
```

4. Copy the [Dockerfile](Dockerfile) and [install.bash](install.bash) file into the directory.

> 5. Submit the build to Cloud Build: `gcloud builds submit --timeout "25m"  --config cloudbuild.yaml .`

Reviewer: Should probably explicitly specify `--project`.
